apis/compute/v1beta1/computefirewallpolicyrule_reference.go

```diff
@@ -0,0 +1,167 @@
+func NewComputeFirewallPolicyRuleRef(ctx context.Context, reader client.Reader, obj *ComputeFirewallPolicyRule) (*ComputeFirewallPolicyRuleRef, error) {
+	id := &ComputeFirewallPolicyRuleRef{}
+
+	firewallPolicyRef, err := refsv1beta1.ResolveComputeFirewallPolicy(ctx, reader, obj, obj.Spec.FirewallPolicyRef)
+	if err != nil {
+		return nil, err
+	}
+	firewallPolicy := firewallPolicyRef.External
+	if firewallPolicy == "" {
+		return nil, fmt.Errorf("cannot resolve firewallPolicy")
+	}
+
+	id.parent = &ComputeFirewallPolicyRuleParent{FirewallPolicy: firewallPolicy}
+
+	// Get priority. Priority is a required field
+	priority := strconv.Itoa(int(obj.Spec.Priority))
+
+	// Use approved External
+	externalRef := valueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id.External = asComputeFirewallPolicyRuleExternal(id.parent, priority)
+		return id, nil
+	}
+
+	// Validate desired with actual
+	actualParent, actualPriority, err := parseComputeFirewallPolicyRuleExternal(externalRef)
+	if err != nil {
+		return nil, err
+	}
+	if actualParent.FirewallPolicy != firewallPolicy {
+		return nil, fmt.Errorf("spec.firewallPolicyRef changed, expect %s, got %s", actualParent.FirewallPolicy, firewallPolicy)
+	}
+	if actualPriority != priority {
+		return nil, fmt.Errorf("cannot reset `spec.priority` to %s, since it has already assigned to %s",
+			priority, actualPriority)
+	}
+	id.External = externalRef
+	id.parent = &ComputeFirewallPolicyRuleParent{FirewallPolicy: firewallPolicy}
+	return id, nil
+}
+
+func (r *ComputeFirewallPolicyRuleRef) Parent() (*ComputeFirewallPolicyRuleParent, error) {
+	if r.parent != nil {
+		return r.parent, nil
+	}
+	if r.External != "" {
+		parent, _, err := parseComputeFirewallPolicyRuleExternal(r.External)
+		if err != nil {
+			return nil, err
+		}
+		return parent, nil
+	}
+	return nil, fmt.Errorf("ComputeFirewallPolicyRule not initialized from `NewComputeFirewallPolicyRuleRef` or `NormalizedExternal`")
+}
+
+type ComputeFirewallPolicyRuleParent struct {
+	FirewallPolicy string
+}
+
+func (p *ComputeFirewallPolicyRuleParent) String() string {
+	return "locations/global/firewallPolicies/" + p.FirewallPolicy
+}
+
+func asComputeFirewallPolicyRuleExternal(parent *ComputeFirewallPolicyRuleParent, priority string) (external string) {
+	return parent.String() + "/rules/" + priority
+}
+
+func parseComputeFirewallPolicyRuleExternal(external string) (parent *ComputeFirewallPolicyRuleParent, priority string, err error) {
+	tokens := strings.Split(external, "/")
+	if len(tokens) != 6 || tokens[0] != "locations" || tokens[2] != "firewallPolicies" || tokens[4] != "rules" {
+		return nil, "", fmt.Errorf("format of ComputeFirewallPolicyRule external=%q was not known (use firewallPolicies/<firewallPolicy>/rules/<priority>)", external)
+	}
+	parent = &ComputeFirewallPolicyRuleParent{
+		FirewallPolicy: tokens[3],
+	}
+	priority = tokens[5]
```

Comment:
Could you add a validation check for priority (int), so that users know how to fix if given an invalid external ref.